conversation_reconstruction_dataflow/conversation_reconstruction/dataflow_main.py

```diff
@@ -48,83 +50,126 @@ def run(known_args, pipeline_args):
+  join_query = " LEFT JOIN (SELECT ps_tmp.* FROM %s as ps_tmp INNER JOIN (SELECT MAX(rev_id) as max_rid, page_id FROM %s GROUP BY page_id) as tmp ON tmp.page_id == ps_tmp.page_id and ps_tmp.rev_id == tmp.max_rid) as page_state ON ingested.page_id == page_state.ps_tmp.page_id"%(known_args.input_page_state_table,known_args.input_page_state_table)
+  condition_query = " WHERE week==%d and year==%d ORDER BY rev_id_in_int"%(known_args.week, known_args.year)
+  read_query = get_max_query + get_latest_query + join_query + condition_query
   with beam.Pipeline(options=pipeline_options) as p:
 
     # Read the text file[pattern] into a PCollection.
-    filenames = (p | beam.io.Read(beam.io.BigQuerySource(query='SELECT UNIQUE(page_id) as page_id FROM [%s]'%known_args.input_table, validate=True)) 
-                   | beam.ParDo(ReconstructConversation())
-                   | beam.io.Write(bigquery_io.BigQuerySink(known_args.output_table, schema=known_args.output_schema, validate=True)))
-
-class ReconstructConversation(beam.DoFn):
-  def process(self, row):
-
-    return
-    input_table = "wikidetox_conversations.test_page_3_issue21" 
+    reconstruction_results, page_states = (p | beam.io.Read(beam.io.BigQuerySource(query=read_query, validate=True)) 
+                   | beam.Map(lambda x: (x['ingested_page_id'], x))
+                   | beam.GroupByKey()
+                   | beam.ParDo(ReconstructConversation()).with_outputs('page_states', main = 'reconstruction_results'))
+    page_states | "WritePageStates" >> beam.io.Write(bigquery_io.BigQuerySink(known_args.page_states_output_table, schema=known_args.page_states_output_schema, write_disposition='WRITE_APPEND', validate=True))
+    reconstruction_results | "WriteReconstructedResults" >> beam.io.Write(bigquery_io.BigQuerySink(known_args.output_table, schema=known_args.output_schema, validate=True))
 
-    logging.info('USERLOG: Work start')
-    page_id = row['page_id']
-    logging.info('Read page_id: %s'%page_id)
 
-    client = bigquery_op.Client(project='wikidetox-viz')
-    query = ("SELECT rev_id FROM %s WHERE page_id = \"%s\" ORDER BY timestamp"%(input_table, page_id))
-    query_job = client.run_sync_query(query)
-    query_job.run()
-    rev_ids = []
-
-    for row in query_job.rows:
-        rev_ids.append(row[0])
-    logging.info('Retrieved revision list: %d revisions'%len(rev_ids))
-
-    construction_cmd = ['python2', '-m', 'construct_utils.run_constructor', '--table', input_table, '--revisions', json.dumps(rev_ids)]
-    construct_proc = subprocess.Popen(construction_cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, bufsize = 4096)
+class ReconstructConversation(beam.DoFn):
+  def QueryResult2json(self, query_res):
```

Comment:
I think a bit a comment might be helpful here... what is query_res, and why are we doing this character slicing?